internal/util-logging/src/main/scala/sbt/internal/util/ManagedLogger.scala

```diff
@@ -35,29 +35,53 @@ class ManagedLogger(
     )
   }
 
-  private lazy val SuccessEventTag = scala.reflect.runtime.universe.typeTag[SuccessEvent]
   // send special event for success since it's not a real log level
   override def success(message: => String): Unit = {
     if (terminal.fold(true)(_.isSuccessEnabled)) {
       infoEvent[SuccessEvent](SuccessEvent(message))(
         implicitly[JsonFormat[SuccessEvent]],
-        SuccessEventTag
+        StringTypeTag.fast[SuccessEvent],
       )
     }
   }
 
-  def registerStringCodec[A: ShowLines: TypeTag]: Unit = {
+  @deprecated("Use macro-powered StringTypeTag.fast instead", "1.4.0")
+  def registerStringCodec[A](
+      s: ShowLines[A],
+      tt: scala.reflect.runtime.universe.TypeTag[A]
+  ): Unit = {
+    LogExchange.registerStringCodec[A](s, tt)
+  }
+  def registerStringCodec[A: ShowLines: StringTypeTag]: Unit = {
     LogExchange.registerStringCodec[A]
   }
 
-  final def debugEvent[A: JsonFormat: TypeTag](event: => A): Unit = logEvent(Level.Debug, event)
-  final def infoEvent[A: JsonFormat: TypeTag](event: => A): Unit = logEvent(Level.Info, event)
-  final def warnEvent[A: JsonFormat: TypeTag](event: => A): Unit = logEvent(Level.Warn, event)
-  final def errorEvent[A: JsonFormat: TypeTag](event: => A): Unit = logEvent(Level.Error, event)
-  def logEvent[A: JsonFormat: TypeTag](level: Level.Value, event: => A): Unit = {
+  @deprecated("Use macro-powered StringTypeTag.fast instead", "1.4.0")
+  final def debugEvent[A](event: => A, f: JsonFormat[A], t: TypeTag[A]): Unit =
+    debugEvent(event)(f, StringTypeTag.apply(t))
+  @deprecated("Use macro-powered StringTypeTag.fast instead", "1.4.0")
+  final def infoEvent[A](event: => A, f: JsonFormat[A], t: TypeTag[A]): Unit =
+    infoEvent(event)(f, StringTypeTag.apply(t))
+  @deprecated("Use macro-powered StringTypeTag.fast instead", "1.4.0")
+  final def warnEvent[A](event: => A, f: JsonFormat[A], t: TypeTag[A]): Unit =
+    warnEvent(event)(f, StringTypeTag.apply(t))
+  @deprecated("Use macro-powered StringTypeTag.fast instead", "1.4.0")
+  final def errorEvent[A](event: => A, f: JsonFormat[A], t: TypeTag[A]): Unit =
+    errorEvent(event)(f, StringTypeTag.apply(t))
```

Comment:
As mentioned in https://github.com/sbt/sbt/pull/5712#discussion_r462737410, `debugEvent[A: JsonFormat: TypeTag](event: => A): Unit` is effectively public, so I don't think we can remove the method.

Reply:
I didn't remove any methods. This passed mima. `debugEvent[A: JsonFormat: TypeTag](event: => A)` desugars to `debugEvent[A](event: => A)(implicit f: JsonFormat[A], t: TypeTag[A])` which is further desugared to a method with the signature `debugEvent[A](event: A, f: JsonFormat[A], t: TypeTag[A])`. For mima purposes, all three versions are the same so binary compatibility is preserved. What this signature change does is allow an automatic upgrade to the new version, `debugEvent[A: JsonFormat: StringTypeTag]` without having to actually change their source code. Builds or plugins compiled for scala 1.4.0 will use the `StringTypeTag` variant but plugins compiled for earlier versions of sbt that use the TypeTag variant will still work with 1.4.0.

Reply:
> which is further desugared to a method with the signature `debugEvent[A](event: A, f: JsonFormat[A], t: TypeTag[A])`.

I didn't know about this part. ok then I stand corrected.